mmv1/third_party/terraform/services/kms/data_source_google_kms_key_rings.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+package kms
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+	transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+)
+
+func DataSourceGoogleKmsKeyRings() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGoogleKmsKeyRingsRead,
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `Project ID of the project.`,
+			},
+			"location": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The canonical id for the location. For example: "us-east1".`,
+			},
+			"filter": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: `
+					The filter argument is used to add a filter query parameter that limits which keys are retrieved by the data source: ?filter={{filter}}.
+					Example values:
+					
+					* "name:my-key-" will retrieve key rings that contain "my-key-" anywhere in their name. Note: names take the form projects/{{project}}/locations/{{location}}/keyRings/{{keyRing}}.
+					* "name=projects/my-project/locations/global/keyRings/my-key-ring" will only retrieve a key ring with that exact name.
+					
+					[See the documentation about using filters](https://cloud.google.com/kms/docs/sorting-and-filtering)
+				`,
+			},
+			"key_rings": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "A list of all the retrieved key rings",
+				Elem: &schema.Resource{
+					// schema isn't used from resource_kms_key_ring due to having project and location fields which are empty when grabbed in a list.
```

Comment:
You can pass those (known) values into your flattener `flattenKMSKeyRingsList` and set the project and location on each item in the list, and that way you can have rich data for each key ring being returned. What do you think?

Reply:
I attempted this but still got empty `project, location` values for each keyring. I can attempt this again assuming i did something wrong.

Reply:
> You can pass those (known) values into your flattener `flattenKMSKeyRingsList` and set the project and location on each item in the list, and that way you can have rich data for each key ring being returned. What do you think?

I was thinking about this also but it seems like extra info that's not really useful since `project` and `location` must be set in order to use the data source. Would show a list of keyRings with all the same locations and projects that were initially set.

An example of what this would look like if a user were to set both `project` and `location` fields while also including them in each `keyRing`
```hcl
.....
          + {
              + id       = "projects/hc-terraform-testing/locations/global/keyRings/tf-test-keyring-examplekipwkpbmlh"
              + location = "global"
              + name     = "tf-test-keyring-examplekipwkpbmlh"
              + project  = "hc-terraform-testing"
            },
          + {
              + id       = "projects/hc-terraform-testing/locations/global/keyRings/tf-test-keyring-exampleksrf4q2e41"
              + location = "global"
              + name     = "tf-test-keyring-exampleksrf4q2e41"
              + project  = "hc-terraform-testing"
            },
          + {
              + id       = "projects/hc-terraform-testing/locations/global/keyRings/tf-test-keyring-exampleml7ke81gcw"
              + location = "global"
              + name     = "tf-test-keyring-exampleml7ke81gcw"
              + project  = "hc-terraform-testing"
            },
        ]
      + location  = "global"
      + project   = "hc-terraform-testing"
 ```

Reply:
@BBBmau Oh wait, I can see that the API response when listing crypto key rings [only includes names and createTime values](https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings#KeyRing), so there isn't other data to put in the schema of each list item. In that case keeping just name and createTime is fine!

Reply:
I just realised that the API only returns name and id, and the crypto key ring resource doesn't have additional fields to fill. In that case I think it's fine to only include id, name, and createTime using a 'custom schema'. I was thinking we should include project and location values if the crypto key ring resource's schema was more substantial

Reply:
Agreed that adding location and project to each entry doesn't make sense in that case.